knimin/lib/data_access.py

```diff
@@ -823,9 +823,112 @@ def pulldown(self, barcodes, blanks):
                                              for h in headers]))
             metadata[survey] = '\n'.join(survey_md).encode('utf-8')
 
-        failures = sorted(set(barcodes) - barcodes_seen)
+        failures = set(barcodes) - barcodes_seen
 
-        return metadata, failures
+        return metadata, self._explain_pulldown_failures(failures)
+
+    def _explain_pulldown_failures(self, barcodes):
+        """Builds failure reason list for barcodes passed
+
+        Parameters
+        ----------
+        barcodes : list of str
+            Barcodes to explain failure for
+
+        Returns: list of tuples of str
+            list in the form [(barcode, reason), (barcode, reason), ...]
+        """
+        # if empty list passed, don't touch database
+        if len(barcodes) == 0:
+            return []
+
+        def update_failure(sql, all_bc, reason):
+            hold = [x[0] for x in
+                    self._con.execute_fetchall(sql, [tuple(all_bc)])]
+            remaining_bc = all_bc.difference(hold)
+            return [(bc, reason) for bc in hold], remaining_bc
+
+        failure_info = []
+        current_barcodes = set(barcodes)
+
+        # TEST ORDER HERE MATTERS! Assumptions made based on filtering of
```

Comment:
Is it better to remove the barcodes and only flag a single error, or to flag all errors associated with a barcode? (Just asking, I do not know what is best here)